### PR TITLE
New load model

### LINKED ISSUE #45: [ENH] New module load process
# Description  
Right now GmodDotNet is very opinionated about module load process, especially on client: clientside modules loaded only after the specific lua event, all modules are loaded simultaneously, etc. It sets limitations on usage and inconsistent with behavior of native Garry's Mod modules.

# Solution proposal

Modules will be loaded and unloaded by name with `dotnet_load` and `dotnet_unload` Lua functions. Code signing functionality will be provided as a separate managed module.

# TODO
- [x] Implement new module load/unload functionality.

- [ ] Develop and ship an optional managed module to provide code signing functionality.

### LINKED ISSUE #47: [ENH] New way to register managed delegates with Garry's Mod Lua engine.
# Description 

Current way of pushing managed delegate to Gmod's Lua engine consists of creating native callable wrapper for delegate with `Marshal` class, optionally applying safe wrapper, and pushing resulting native function pointer to lua stack. This process is suboptimal, since end users can't distinguish native functions from managed in lua code.

# TODO

Some experiments need to be done to get a finale solution, but current sketch is following:

- [x] Introduce new Lua type for callable objects to represent managed functions.

- [x] Replace `ILua.PushCFunction(CFuncManagedDelegate)` with something like `ILua.PushManagedDelegate(Func<IntPtr, int>)`.

- [x] Eliminate need for manual delegate reference retention by working with `__gc` metamethod of custom Lua type.

### LINKED ISSUE #57: [ENH] Allow native dependencies to be unloaded on module unload
# Overview

Current `GmodNetModuleAssemblyLoadContext` implementation is able to unload only managed parts of GmodDotNet modules. Thus if module brings native dependencies the game process retains reference to them until its termination, which effectively means that such modules are not really hot-reloadable.

# Possible solution

Each instance of `GmodNetModuleAssemblyLoadContext` should keep track of all native libraries requested as dependencies and free their corresponding OS references with `NativeLibrary.Free`.

### LINKED ISSUE #66: [ENH] Update bundled .NET Runtime to version 5.0.0 RTM

### PR DESCRIPTION
**Pull request by GlebChili**
Closes linked issues

### CODE REVIEW

**Comment by GlebChili** [comment]
@AaronJThompson Ready for review.

File: gm_dotnet_managed/GmodNET.API/GmodInterop.cs [hunk 1]
```diff
@@ -25,5 +25,23 @@ public static ILua GetLuaFromState(IntPtr lua_state)
...
+        /// Get an upvlaue pseudo-index of the Lua closure.
+        /// </summary>
+        /// <param name="upvalue">A relative index of the upvalue.</param>
+        /// <param name="managed_offset">Use upvalue offset for managed closures</param>
+        /// <returns>A pseudo-index to access upvalue.</returns>
+        public static int GetUpvalueIndex(byte upvalue, bool managed_offset = true)
+        {
+            if(managed_offset)
+            {
+                return (-10003 - upvalue);
```
Comment (AaronJThompson) [inline gm_dotnet_managed/GmodNET.API/GmodInterop.cs:39]: Can I just get a quick explanation of why managed upvalues are indexed differently?

**Review by AaronJThompson — COMMENTED** [review]
Honestly, this was a great PR. Not really much to comment on, the code looks great. Will approve after comments resolved

File: gm_dotnet_managed/GmodNET.API/GmodInterop.cs (hunk 1 above)
Comment (GlebChili) [reply gm_dotnet_managed/GmodNET.API/GmodInterop.cs:39]: @AaronJThompson Whenever we push managed function, we actually push a function closure with first upvalue being a GCHandle integer representation (see https://github.com/GmodNET/GmodDotNet/blob/4dfb83d26104906fd2ae56c332f27f59ac4980c0/gm_dotnet_managed/GmodNET/Lua.cs#L624). Thus, if we have a managed closure, the real first upvalue will be GCHandle. If we want access first practical upvalue we indeed need to access second upvalue. Therefore we have this -1 offset.

**Review by AaronJThompson — APPROVED** [review]
Awesome work! Again, great PR. Looking forward to working with the new model! Approved

**Pull request merged**